src/io/tests/test_hdf_manual_reorder.cpp

```diff
@@ -0,0 +1,212 @@
+  REQUIRE(read_tensor.size() == v.size());
+  for (int i = 0; i < 3; i++)
+  {
+    for (int j = 0; j < 3; j++)
+    {
+      int pos = i*3+j;
+      REQUIRE(v[pos] == read_tensor(i,j));
+    }
+  }
+  hd2.close();
+}
+*/
+
+TEST_CASE("hdf_write_reorder_with_matrix", "[hdf]")
+{
+  hdf_archive hd;
+  bool okay = hd.create("test_write_matrix_reorder.hdf");
+  REQUIRE(okay);
+
+  Vector<double> v(6);
+  v[0] = 0.0;
+  v[1] = 0.1;
+  v[2] = 0.2;
+  v[3] = 1.0;
+  v[4] = 1.1;
+  v[5] = 1.2;
+
+  Matrix<double> matrix_view(v.data(), 2, 3);
+  hd.write(matrix_view, "matrix_from_vector");
+
+  hdf_archive hd2;
+  hd2.open("test_write_matrix_reorder.hdf");
+
+  Matrix<double> m(2, 3);
+  hd2.read(m, "matrix_from_vector");
+
+  REQUIRE(m(0, 0) == Approx(v[0]));
+  REQUIRE(m(0, 1) == Approx(v[1]));
+  REQUIRE(m(0, 2) == Approx(v[2]));
+  REQUIRE(m(1, 0) == Approx(v[3]));
+  REQUIRE(m(1, 1) == Approx(v[4]));
+  REQUIRE(m(1, 2) == Approx(v[5]));
+}
+
+
+// do this in three ways.
+// 1. relying on my overloading in std::vector (do not perfer this, too specific)
+// 2. rely on my understanding of hyperslab_proxy (seems to require you to know
+//          too much about the dimensions in the dataset and is prone to user error)
+// 3. use hyperslab_proxy to handle the back end api, but specify the slice to choose
+//          with a vector and query the dataset to figure out the other needed
+//          parts.  better than #1 because it does not specially call out vector
+TEST_CASE("hdf_read_partial", "[hdf]")
+{
+  hdf_archive hd;
+  bool okay = hd.create("test_read_partial.hdf");
+  REQUIRE(okay);
+
+  Matrix<double> allData(3, 4);
+  for (int i = 0; i < 3; i++)
+  {
+    for (int j = 0; j < 4; j++)
+    {
+      allData(i, j) = i + j * 0.1;
+    }
+  }
+
+  hd.write(allData, "matrix");
+  hd.close();
+
+  hdf_archive hd2;
+  okay = hd2.open("test_read_partial.hdf");
+  REQUIRE(okay);
+
+  // method 1 (relying on vector in hdf_stl)
+  vector<int> readSpec{1, -1};
+  vector<double> readBuffer(4);
+
+  hd2.read(readBuffer, readSpec, "matrix");
+  for (int i = 0; i < 4; i++)
+  {
+    REQUIRE(readBuffer[i] == Approx(allData(1, i)));
+  }
+
+  readSpec[0] = -1;
+  readSpec[1] = 2;
+  hd2.read(readBuffer, readSpec, "matrix");
+  for (int i = 0; i < 3; i++)
+  {
+    REQUIRE(readBuffer[i] == Approx(allData(i, 2)));
+  }
+
+  readSpec[0] = 2;
+  readSpec[1] = 0;
+  hd2.read(readBuffer, readSpec, "matrix");
+  REQUIRE(readBuffer[0] == Approx(allData(2, 0)));
+
+  // method 2 (direct utilization of hyperslab_proxy in client code)
+  Matrix<double> outbuffer1(1, 4);
+  Matrix<double> outbuffer2(3, 1);
+  Matrix<double> outbuffer3(1, 1);
+
+  TinyVector<int, 2> dims_unused;
+  dims_unused[0] = 1;
+  dims_unused[1] = 2;
+  TinyVector<int, 2> dims_local;
+  dims_local[0] = 1;
+  dims_local[1] = 4;
+  TinyVector<int, 2> offsets;
+  offsets[0] = 1;
+  offsets[1] = 0;
+
+  hyperslab_proxy<Matrix<double>, 7> pxy1(outbuffer1, dims_unused, dims_local, offsets);
```

Comment:
We can do smarter defaults for dims_unused, offsets, dims_locals instead of forcing users to specify all of them.

When using hyperslab, I consider
(offset)
(offset, dim_local)
(offset, dim_local, dim_all)
all useful for reading
but only the last one is valid for writing.